### ISSUE TITLE
Filtering is terribly slow

### ISSUE THREAD

**Issue by balta2ar** [Shougo/unite.vim#848]
Unfortunately I'm having a bad user experience when using `file_rec/async`. To take it to an extreme, I run it in the root of the linux kernel tree, that's about 50K files: `Unite file_rec/async`. I do not use any fancy fuzzy matchers (well, I think, I do not overwrite any default matcher options), or anything other than a default setup. In fact, it's not find that is slow:

```
linux-3.19 $ time find -type f | wc -l                                                    
48425
noglob find -type f  0.03s user 0.07s system 98% cpu 0.105 total
```

If I understand it right, Unite caches the results of find. If I run `file_rec/async` for first time on linux tree, then close the Unite window, and then run `file_rec/async` again, it reacts too slow (it opens in less than a second, but very close it). Also, when I type something, it reacts equally painfully slowly.

For comparison, simple grep runs relatively fast on this moderately sized data:

```
linux-3.19 $ time find -type f | grep radeon/cik | wc
      6       6     318
noglob find -type f  0.03s user 0.07s system 94% cpu 0.109 total
grep -i --color=always radeon/cik  0.02s user 0.00s system 18% cpu 0.109 total
wc  0.00s user 0.00s system 0% cpu 0.110 total
```

I tried to dig into the source code trying to find out who's making it slow. Before executing `file_rec/async` I used profile as follows: `:profile start vim.log | profile file * | profile func *`. I noticed that filtering function is taking quite some time: `function! unite#filters#lua_filter_patterns(candidates, patterns, whites)` (I have +lua). Then I put `return a:candidates` as the first line in it but still it was terribly slow.

Here are the profile results with `unite#filters#lua_filter_patterns` doing nothing (simply returning original candidates). I only started `:Unite file_rec/async`, didn't type anything and quit vim.

```
FUNCTION  522()
Called 1 time
Total time:   0.468213
 Self time:   0.468213

count  total (s)   self (s)
    1              0.000003   if stridx(a:context.input, '.') >= 0
                                return unite#filters#filter_matcher( a:candidates, '', a:context)
                              endif
                            
                              "return a:candidates
    1              0.468206   return filter(a:candidates, " get(v:val, 'action__path', v:val.word) !~    '\\%(^\\|/\\)\\.[^/]*/\\?$'")

FUNCTIONS SORTED ON TOTAL TIME
count  total (s)   self (s)  function
    1   1.075987   0.000351  <SNR>109_call_unite_empty()
    1   1.071344   0.000457  unite#start()
    1   1.069877   0.001084  unite#start#standard()
    1   1.007665   0.000514  unite#candidates#_recache()
    1   1.005815   0.000301  <SNR>180_recache_candidates_loop()
    2   0.560795   0.000130  unite#helper#call_source_filters()
    5   0.560665   0.000097  unite#helper#call_filter()
    1   0.468213             522()
    1   0.423415   0.028123  <SNR>180_get_source_candidates()
    1   0.389082   0.030104  507()
    1   0.345626   0.207106  <SNR>176_init_continuation()
    1   0.118036             unite#helper#paths2candidates()
    1   0.078722   0.000077  516()
    1   0.078617   0.078614  unite#filters#converter_relative_word#lua()
    8   0.045314   0.003302  <SNR>163__import()
    5   0.038448   0.000286  <SNR>163_import()
    9   0.031766   0.017227  unite#init#_default_scripts()
    1   0.030565   0.000552  unite#init#_unite_buffer()
    1   0.024920   0.000811  unite#init#_current_unite()
    5   0.023946   0.000083  <SNR>120_on_window_changed()

FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
    1              0.468213  522()
    1   0.345626   0.207106  <SNR>176_init_continuation()
    1              0.118036  unite#helper#paths2candidates()
    1   0.078617   0.078614  unite#filters#converter_relative_word#lua()
    1   0.389082   0.030104  507()
    1   0.423415   0.028123  <SNR>180_get_source_candidates()
    1   0.020278   0.020199  <SNR>177_readfile()
    9   0.031766   0.017227  unite#init#_default_scripts()
    8   0.019838   0.012000  <SNR>163__build_module()
   11   0.018173   0.009906  <SNR>163__scripts()
   14              0.009884  <SNR>163__redir()
  240              0.007611  <SNR>150_get_syn()
  162              0.007030  airline#highlighter#exec()
    3   0.017505   0.005467  airline#highlighter#highlight()
   10              0.005102  <SNR>163__runtime_files()
   87              0.004993  <SNR>198_is_file_exist()
    1   0.006192   0.004907  <SNR>180_ignore_candidates()
   46   0.016926   0.004407  425()
  120   0.013560   0.004223  airline#highlighter#get_highlight()
    2              0.003735  <SNR>198_uniq_by()
```
(full log is here: http://sprunge.us/GGND)

I know that VimL is known to be rather slow language, but maybe still there is something that could be done? Maybe reduce some excessive caching or `deepcopy`ing or implement filtering in `C`, I don't know.

Could there be some kind of "filtering backend", sort of completion backend that does the filtering blazingly fast and returns only very small potion of result? At any given point of time there's still a limit to the number lines that we can show on the screen anyway, plus it looks like `file_rec/async` is also limiting the output: `file_rec/async(50/48436)` -- I mean the first number = `50`.

What do you think about all that? Is it possible to speed things up?

**Comment by Shougo** [comment]
You should change `g:unite_redraw_hold_candidates` value.

But I will improve matcher_hidden_files.

**Comment by balta2ar** [comment]
Thanks, setting it to 50K made it much better (however, delay is still noticeable).

**Comment by Shougo** [comment]
Optimized.

**Comment by balta2ar** [comment]
Thanks!

One question, though. Again, I test on linux kernel sources. I have this in my config:

```
let g:unite_redraw_hold_candidates = 50000
call unite#custom#source('file_rec/async', 'matchers',
  \ ['converter_relative_word', 'matcher_fuzzy'])
```

It is reasonable fast at this moment. However, as soon as I add `'matcher_project_ignore_files'` after `'converter_relative_word'` it becomes several orders of magnitude slower and it literally takes tens of seconds to open file_rec/async and filter results. Is it supposed to be that slow and nothing can be done about it?

**Comment by balta2ar** [comment]
I've looked at `unite#filters#lua_filter_patterns` and added printing the number of arguments:

Matcher | len(a:candidates) | len(a:patterns) | len(a:whites)
------- | --- | --- | ---
matcher_fuzzy | 6725618 | 215 | 2
matcher_project_ignore_files | 6725618 | 31288 | 224

Judging by three nested `for`s in `lua_filter_patterns` it makes it `O(n^3)`, no wonder it becomes slower as number of patterns and whites grows. Please correct me if I'm wrong and misinterpreted the implementation. This is just my speculations as I didn't run the profiler to prove my assumption.

Where can I read about this function `string.find` in lua code and see its implementation?

**Comment by Shougo** [comment]
You should use this:

```vim
call unite#custom#source('file_rec/async', 'matchers',
  \ ['converter_relative_word', 'matcher_fuzzy', 'matcher_project_ignore_files'])
```

**Comment by Shougo** [comment]
> I've looked at unite#filters#lua_filter_patterns and added printing the number of arguments:

Why?
See the below code:

```vim
function! s:matcher.filter(candidates, context) "{{{
  if stridx(a:context.input, '.') >= 0
    return a:candidates
  endif

  return unite#util#has_lua() ?
        \ unite#filters#lua_filter_patterns(a:candidates,
        \   ['^%.[^/]*/?$', '/%.[^/]*/?$'], []) :
        \ filter(a:candidates, "
        \   has_key(v:val, 'action__path')
        \    && v:val.action__path !~ '\\%(^\\|/\\)\\.[^/]*/\\?$'")
endfunction"}}}
```

`len(a:patterns)` = 2 and `len(a:whites)` = 0.   Your result is wrong.

**Comment by Shougo** [comment]
> Judging by three nested fors in lua_filter_patterns it makes it O(n^3), no wonder it becomes slower as number of patterns and whites grows. Please correct me if I'm wrong and misinterpreted the implementation. This is just my speculations as I didn't run the profiler to prove my assumption.

No.  You can read the code?  Yes, it is three nested loop.  But it is candidates-patterns-whites loop.
Not O(n^3) order.

**Comment by Shougo** [comment]
> It is reasonably fast at this moment. However, as soon as I add 'matcher_project_ignore_files' after 'converter_relative_word' it becomes several orders of magnitude slower and it literally takes tens of seconds to open file_rec/async and filter results. Is it supposed to be that slow and nothing can be done about it?

OK.  I will check the performance.

**Comment by Shougo** [comment]
> I've looked at unite#filters#lua_filter_patterns and added printing the number of arguments:

Your files are 48425 candidates.  But it is too different...

**Comment by balta2ar** [comment]
> len(a:patterns) = 2 and len(a:whites) = 0. Your result is wrong.

No no, the code you've shown is from `matcher_hide_hidden_files.vim`, and I'm talking about `matcher_project_ignore_files.vim`

```
call unite#custom#source('file_rec/async', 'matchers',
  \ ['converter_relative_word', 'matcher_fuzzy', 'matcher_project_ignore_files'])
```
OK, I've changed the order as you adviced, and filtering when Unite is open is not quite resoponsive but tolerable. But it is still very slow at opening the Unite buffer for the first time.

> No. You can read the code? Yes, it is three nested loop. But it is candidates-patterns-whites loop.
Not O(n^3) order.

Yes, I'm sorry, my conclusion was too rash, it's not `O(n^3)` but rather `O(n * m * k)`. I was just thinking about low-hanging fruits and at first it looked like a good candidate for Aho–Corasick, but later I realized that patterns and whites contain regular expressions.

**Comment by balta2ar** [comment]
For the record, I'm using files from this archive: https://www.kernel.org/pub/linux/kernel/v3.x/linux-3.19.tar.xz

Ok, it looks like my numbers are indeed trash because I convert candidates to string before calling len. But it really doesn't matter because what matters is performance and it's low. The correct table should look like this:

Matcher | len(a:candidates) | len(a:patterns) | len(a:whites)
------- | --- | --- | ---
matcher_fuzzy | 48435 | 15 | 0
matcher_project_ignore_files | 48435 | 538 | 4

My diff: http://sprunge.us/PHZM?diff. Please check the performance of `matcher_project_ignore_files`.

**Comment by Shougo** [comment]
OK.  I get it.  It is feature.
You should use "required_pattern_length" option.

```vim
call unite#custom#source('file_rec/async', 'required_pattern_length', 2)
```

**Comment by balta2ar** [comment]
I'm not sure I understand how this is related. It looks like `required_pattern_length` just delays the moment when filtering starts, it doesn't solve the performance problems.

How does `git status` and `hg status` commands work? How do they make it fast? For example, here are timing results of `git status` on linux tree:

``` bash
linux → for _ in {1..10}; /usr/bin/time -f "%e" git status > /dev/null 
0.18
0.15
0.14
0.13
0.14
0.13
0.13
0.13
0.15
0.14
``` 

This is always below 200ms which is a decent delay. Yes, git is implemented in pure C, but maybe there is some optimization that can be done in Unite to achieve similar speed?

**Comment by Shougo** [comment]
> I'm not sure I understand how this is related. It looks like required_pattern_length just delays the moment when filtering starts, it doesn't solve the performance problems.

Yes.  But it solves below problem.

> OK, I've changed the order as you adviced, and filtering when Unite is open is not quite resoponsive but tolerable. But it is still very slow at opening the Unite buffer for the first time.

> This is always below 200ms which is a decent delay. Yes, git is implemented in pure C, but maybe there is some optimization that can be done in Unite to achieve similar speed?

It is impossible.  I think 0.18s in C is too slow.

You should use "file_rec/git" source and should not use "matcher_project_ignore_files" in that case.

**Comment by balta2ar** [comment]
> You should use "file_rec/git" source and should not use "matcher_project_ignore_files" in that case.

Yeah, looks like a viable solution. In my project, though, I use hg. I had to use this option:

``` vim
let g:unite_source_rec_async_command = 'hg st -Aqun'
```

By the way, looks like this option does not support piped commands. My first attempt was this:

``` vim
let g:unite_source_rec_async_command = 'hg st -Aqu | cut -c 2-'
```

And it didn't work. This kinda limits to flexibility of the option. Maybe this should be fixed?

Is there any way to set this option per project?

**Comment by Shougo** [comment]
It is feature.
Because, file_rec must set directory argument like this:

```vim
hg st -Aqun {directory}
```

**Comment by Shougo** [comment]
> Is there any way to set this option per project?

Nothing.

> Yeah, looks like a viable solution. In my project, though, I use hg. I had to use this option:

I thought your project was Linux kernel....

**Comment by balta2ar** [comment]
>> Yes, git is implemented in pure C, but maybe there is some optimization that can be done in Unite to achieve similar speed?

> It is impossible. I think 0.18s in C is too slow.

Okay, I've performed some experiments. I don't want to jump into conclusions, but I think it is possible.

**TL;DR: I've managed to reduce `unite#filters#lua_filter_patterns()` execution time from 34.367448 to 1.005875 seconds on linux kernel source tree.**

Before:
```
FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
    2             34.367448  unite#filters#lua_filter_patterns()
    1   1.664487   1.616389  <SNR>185_get_ignore_results()
```

After:
```
FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
    1   1.706482   1.656941  <SNR>185_get_ignore_results()
    2              1.005875  unite#filters#lua_filter_patterns()
```

The trick is to use Google's RE2 regular expression library (http://code.google.com/p/re2/). There is `Set` class (http://code.google.com/p/re2/source/browse/re2/set.h) that can combine multiple regular expressions into one and test the string against all of them at once (instead of testing the string against N expressions sequentially).

I wrote a small C++ program that receives three files: with words, patterns and whites. It combines all `patterns` and `whites` regular expressions into two `Set`s and that gave me the speed improvement. The result of the filtering is saved into another file. `unite#filters#lua_filter_patterns()` saves arguments into files, calls this external program using `os.execute` and reads the results from file when it's done. My current implementation is just a proof-of-concept, it's very crude and so on, but it looks like it proves the point.

Below I show how to reproduce my results. Please try it and see if I missed anything, if it works at all on your side. If it is, this leads me back to the idea that I offered in the first post: what do you think about some fast optimized filtering backend (maybe a plugin like like vim-proc) that can be an optional part of Unite? It's not mandatory, and by default `matcher_project_ignore_files` is slow, but if user enables the plugin, it makes all filtering even faster. Another way to implement this is to make a C library and add Lua interface to it. Well, that's just my thoughts, I guess you know how to do that better.

Please note that there are some problems with current solution, for example filtering is not quite the same. This is because you translate the usual regular expressions into the format suitable for Lua. I disabled some of thess translations, but not all (I didn't want to spend time on that), however results with and without external filter are more or less the same. 

### How to reproduce

My filtering settings:
``` vim
call unite#custom#source('file_rec/async', 'matchers',
 \ ['matcher_default', 'matcher_project_ignore_files'])
```

Steps to reproduce:
``` bash
# Prepare temporary folders
mkdir /tmp/unite
mkdir /tmp/reproduce
cd /tmp/reproduce

# Prepare re2
hg clone https://re2.googlecode.com/hg re2
cd re2
make -j4
cd ..

# Compile external filter
wget -O client.cpp http://sprunge.us/LSRO
g++ client.cpp -o client -g -O -fPIC -lstdc++ -Ire2 -Lre2/obj/so -lpthread -lre2 -x c++ -std=c++0x

# Apply my changes to Unite
cd ~/.vim/bundle/unite.vim
wget -O flt.patch http://sprunge.us/iSiI
git apply flt.patch

# Make sure vim knows where to load re2
cd /tmp/src/linux-3.19
LD_LIBRARY_PATH=/tmp/reproduce/re2/obj/so vim
```

I used this command in vim before calling `Unite file_rec/async` to profile:
``` vim
:profile start vim.log | profile file * | profile func *
```

### Profile results

Before: http://sprunge.us/jFeJ

After: http://sprunge.us/ENgZ

**Comment by Shougo** [comment]
OK.  Thanks.
I think to use "std::regex" is better than R2 library.

The feature should be option.  And it can use other filters (eg: matcher_fuzzy).
Your solution for if_lua, but it can be execute for VimL.

**Comment by balta2ar** [comment]
> I think to use "std::regex" is better than R2 library.

I never used it. Can it combine multiple expressions into one? I guess it's a good idea to compare their perfomance and choose the one which is faster.

**Comment by Shougo** [comment]
> Can it combine multiple expressions into one?

std::regex supports "|" pattern.

> I guess it's a good idea to compare their perfomance and choose the one which is faster.

Standard regex library may slow, but we don't have to other libraries.  I think it is better.

**Comment by balta2ar** [comment]
> Standard regex library may be slow, but we don't have to complie other libraries. I think it is better.

Sorry, I can't agree with you on this. Adding re2 is not that big of hassle, and if its performance is noticably better, I think it's worth using it. For example, YouCompleteMe includes third party libraries and it's perfectly fine. As long as compiling it is as easy as calling `./install`, I'd go with that. Unite can use re2 optionally, if it's available. I'd choose difficulties of compiling optional dependencies rather than having a slow default program.

**Comment by Shougo** [comment]
Please create performance test with Re2 and std::regex.

**Comment by balta2ar** [comment]
Results are poor for std::regex.

```
→ time LD_LIBRARY_PATH=re2/obj/so ./client re2 data/candidates-small.txt data/patterns.txt data/whites.txt result-re.txt
LD_LIBRARY_PATH=re2/obj/so ./client re2 data/candidates-small.txt     0.07s user 0.05s system 91% cpu 0.136 total

→ time LD_LIBRARY_PATH=re2/obj/so ./client std data/candidates-small.txt data/patterns-backslash-hypen.txt data/whites-backslash-hypen.txt result-std.txt
LD_LIBRARY_PATH=re2/obj/so ./client std data/candidates-small.txt     108.48s user 0.05s system 99% cpu 1:48.66 total
```

1 minute and 48 seconds for std::regex. This is even slower than Lua. Here is my repo which you can use to reproduce my results: https://github.com/balta2ar/unite-perf

Ths reason of this huge difference is that `The one significant exception is that RE2 drops support for backreferences` (https://code.google.com/p/re2/). On the other hand, std::regex support backreferences. See this answer: http://stackoverflow.com/a/20943095/258421

Another nice reading regarding re2: http://swtch.com/~rsc/regexp/regexp1.html

**Comment by Shougo** [comment]
OK.
You should create new Vim plugin library uses Google R2 regex.
unite.vim can support it.

**Comment by balta2ar** [comment]
Erm... How do I do that? Could you please provide more technical details or point me to a similar plugin library that does silimar things (e.g. filtering)? I have very little experience with VimL. Where do I start?

**Comment by balta2ar** [comment]
I started with vimproc plugin as it contains an example of how to access C library from VimL. I see vp_stack* functions that parse arguments from given `char *`. I expect my plugin to be called from `unite/filters.vim` with candidates, patterns and whites as arguments. Is that right?

If so, looks like I have to use this vimstack.c to parse vim arguments. Not sure how to get access to condidates list of dicts. Any help is appreciated.

**Comment by Shougo** [comment]
I think the library is executable file like "client".
You don't have to create C library.

**Comment by balta2ar** [comment]
Please elaborate. How do we pass candidates and patterns for filtering? Could maybe implement very simple integration in Unite so that I could see how you expect to use that library?

**Comment by Shougo** [comment]
> How do we pass candidates and patterns for filtering?

You should use the temporary file like this:

http://sprunge.us/iSiI

> Could maybe implement very simple integration in Unite so that I could see how you expect to use that library?

The library calls the command with the arguments simply.

**Comment by tony** [comment]
@Shougo , @balta2ar :   When I was using ctrlp (far snappier than unite.vim is now) I was able to get nice performance boosts from "matchers".

Example of ctrlp matcher (requires +python): https://github.com/FelikZ/ctrlp-py-matcher/tree/master/autoload

Example of a python wrapper for re2: https://github.com/facebook/pyre2

Since https://github.com/google/re2 isn't on package managers yet, a vim plugin would have grab the code and compile it remotely, ostensibly locally, for us to get access to it.

**Comment by tony** [comment]
I think we can collectively agree unite.vim has a speed problem with larger projects.

Ctrlp is handling large projects ok - but it doesn't suit my needs or have anywhere near the flexibility of unite.vim. 

Here is another example of a speedup for a matcher in vim: https://github.com/JazzCore/ctrlp-cmatcher/

``NeoBundle 'JazzCore/ctrlp-cmatcher', { 'build': { 'others': './install.sh' }}``.

Is there an API documentation unite.vim for creating unite.vim filterers/matchers?  Are there any third-party filterers/matchers for unite.vim on gh to look at as an example?

**Comment by tony** [comment]
@Shougo I think I see it now, just by looking at https://github.com/Shougo/unite.vim/tree/master/autoload/unite/filters.

I just realized I wasn't enabling ``luajit`` in my vim build. I'm updating 

@Shougo / @balta2ar Python is really my strong point, one of you may be more up to this: https://github.com/cloudflare/lua-re2. Either of you are interested?

**Comment by Shougo** [comment]
> When I was using ctrlp (far snappier than unite.vim is now) I was able to get nice performance boosts from "matchers".

Yes, unite.vim is slower than because unite.vim has more features than ctrlp.
You can disable it like "matcher_ignore_project_files" and "ignore_globs".

> @Shougo / @balta2ar Python is really my strong point, one of you may be more up to this: https://github.com/cloudflare/lua-re2. Either of you are interested?

I don't want to add the features.  But Pull request is wellcome.

**Comment by balta2ar** [comment]
Thanks for the links, @tony. Lack of lua bindings is not a show stopper to implement this, though, the lack of free time is. I don't think it's necessary to introduce additional libraries in Unite; instead, as @Shougo reasonably suggested, a separate plugin (filter) should be created and Unite should communicate with it using VimScript in a generic and optional way (it requires integration from Unite side, but it seems to be small efforts). That plugin can internally use lua bindings if it needs to.

However I managed to tune (okay, steal yours, in fact, from related issue) my `g:unite_source_rec_async_command` configuration using `ag`, and it pretty much solved the problem because `ag` filters out files specified in `.hgignore` and it does it fast, leaving only a small portion of files for Unite to work with, which is what I needed. I removed `matcher_ignore_project_files` from my settings as I don't seem to need it anymore. Thus it makes me somewhat less motivated to implement that plugin using re2.

**Comment by Shougo** [comment]
@balta2ar I agree with you.

**Comment by Shougo** [comment]
I will improve the performance later...

**Comment by Shougo** [comment]
Unfortunatelly, I cannot fix it.
I created external process for it, but it is too slow.

@balta2ar 's solution is not valid.  Because, your solution breaks candidates information.
It creates "word", "action__path" attributes only.

**Comment by Shougo** [comment]
> Example of ctrlp matcher (requires +python): https://github.com/FelikZ/ctrlp-py-matcher/tree/master/autoload

matcher_fuzzy already uses if_lua.  It is faster than if_python.
So, your if_python solution is not fast.
